Solutions/Marain.TenantManagement.Specs/Mocks/InMemoryTenantProvider.cs

```diff
@@ -0,0 +1,143 @@
+        {
+            ITenant parent = await this.GetTenantAsync(parentTenantId).ConfigureAwait(false);
+
+            var newTenant = new Tenant(this.jsonSerializerSettingsProvider)
+            {
+                Id = Guid.NewGuid().ToString(),
+                Name = name,
+            };
+
+            List<string> childrenList = this.GetChildren(parent.Id);
+            childrenList.Add(newTenant.Id);
+            this.allTenants.Add(new StoredTenant(newTenant, this.jsonSerializerSettingsProvider.Instance));
+
+            return newTenant;
+        }
+
+        public Task DeleteTenantAsync(string tenantId)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        public async Task<TenantCollectionResult> GetChildrenAsync(string tenantId, int limit = 20, string? continuationToken = null)
+        {
+            ITenant parent = await this.GetTenantAsync(tenantId).ConfigureAwait(false);
+
+            List<string> children = this.GetChildren(parent.Id);
+
+            int skip = 0;
+
+            if (!string.IsNullOrEmpty(continuationToken))
+            {
+                skip = int.Parse(continuationToken);
+            }
+
+            IEnumerable<string> tenants = children.Skip(skip).Take(limit);
+
+            continuationToken = tenants.Count() == limit ? (skip + limit).ToString() : null;
+
+            return new TenantCollectionResult(tenants, continuationToken);
+        }
+
+        public Task<ITenant> GetTenantAsync(string tenantId, string? eTag = null)
+        {
+            ITenant? tenant = tenantId == this.Root.Id
+                ? this.Root
+                : this.allTenants.Find(x => x.Id == tenantId)?.Tenant;
+
+            if (tenant == null)
+            {
+                throw new TenantNotFoundException();
+            }
+
+            return Task.FromResult(tenant);
+        }
+
+        public Task<ITenant> UpdateTenantAsync(ITenant tenant)
+        {
+            StoredTenant? currentStoredTenant = this.allTenants.Find(x => x.Id == tenant.Id);
+
+            if (currentStoredTenant == null)
+            {
+                throw new TenantNotFoundException($"Cannot update tenant '{tenant.Name}' with Id '{tenant.Id}' because it has not previously been saved.");
+            }
+
+            currentStoredTenant.Tenant = tenant;
+
+            return Task.FromResult(tenant);
+        }
+
+        public ITenant? GetTenantByName(string name)
+        {
+            return this.allTenants.Find(x => x.Name == name)?.Tenant;
+        }
+
+        public List<string> GetChildren(string parentId)
+        {
+            if (!this.tenantsByParent.TryGetValue(parentId, out List<string>? children))
+            {
+                children = new List<string>();
+                this.tenantsByParent.Add(parentId, children);
+            }
+
+            return children;
+        }
+
+        private class StoredTenant
```

Comment:
Consider adding a comment explaining why we copy the data out of the tenant andinto this instead of storing the tenant objects. (I know you already explained that to me, I just think it'd be worth making the explanation available to others.)